### ISSUE TITLE
AducIotAgent: error while loading shared libraries: libdeliveryoptimization.so.0 during installation

### ISSUE THREAD

**Issue by buentead** [Azure/iot-hub-device-update#344]
Hi 

I'm in the process to install the device update feature on my Raspberry PI 3B. As I couldn't find Ubuntu 18.04 LTS anymore, I installed Ubuntu 20.04 LTS on the device. Then I followed the steps according to the documentation [Device Update Agent Provisioning](https://learn.microsoft.com/en-us/azure/iot-hub-device-update/device-update-agent-provisioning). Of course I replaced 18.04 with 20.04 where applicable.

Everything is fine up to the installation of the device update agent: IoT Edge Runtime installed, device connected to IoT Hub successfully.

Now I experience problems with the installation of the agent.
`sudo apt-get install deviceupdate-agent`

## Expected Behavior
Installation of device update agent should end successful.

## Current Behavior
The installation process fails with the following output (it also showed the same error when I started the installation the 1st time):
```
>> sudo apt-get install deviceupdate-agent
Reading package lists... Done
Building dependency tree       
Reading state information... Done
deviceupdate-agent is already the newest version (1.0.1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] 
Setting up deviceupdate-agent (1.0.1) ...
******************** Running /var/lib/dpkg/info/deviceupdate-agent.postinst configure ***********************
Setting directories and files...
Generate the template configuration file...
Generating the diagnostics configuration file...
Create data dir...
drwxrwx--- 4 adu adu 4096 Jan 18 09:37 /var/lib/adu
drwxrwx--- 2 adu adu 4096 Jan 18 09:37 /var/lib/adu/downloads
Register all reference update content handlers...
/usr/bin/AducIotAgent: error while loading shared libraries: libdeliveryoptimization.so.0: cannot open shared object file: No such file or directory
dpkg: error processing package deviceupdate-agent (--configure):
 installed deviceupdate-agent package post-installation script subprocess returned error exit status 127
Errors were encountered while processing:
 deviceupdate-agent
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

## Steps to Reproduce
I strictly followed the steps described in [Device Update Agent Provisioning](https://learn.microsoft.com/en-us/azure/iot-hub-device-update/device-update-agent-provisioning).

### Device Information
* Host OS: Ubuntu 20.04.5 LTS (Focal Fossa) 
* Architecture: arm64
* Provisioning Method: NA

## Logs
No logs are available in "/var/log/adu/"

## Additional Information
I used Raspberry Pi Imager v1.7.3 to create the SD card. From there I selected the Ubuntu version 20.04.5 LTS (64-bit) for installation. This build of the OS was used on the device.

**Comment by jelalanne** [comment]
You might have a try with these 18.04 images :)

https://cdimage.ubuntu.com/releases/18.04/release/ 

But same problem as yours, I was looking for an answer too. Funny thing is that I figured this out once, but couldn't remember how I did it.

**Comment by jw-msft** [comment]
First try installing arm64 packages for DeliveryOptimization(DO) in this release tarball:
https://github.com/microsoft/do-client/releases/download/v1.0.0/ubuntu2004_arm64-packages.tar

```sh
wget https://github.com/microsoft/do-client/releases/download/v1.0.0/ubuntu2004_arm64-packages.tar
tar -xvf ubuntu2004_arm64-packages.tar
apt install deliveryoptimization-agent_1.0.0_arm64.deb libdeliveryoptimization_1.0.0_arm64.deb

# if you want to proxy apt traffic through DO
# apt deliveryoptimization-plugin-apt_0.5.1_arm64.deb

# if doing DO development
# apt install libdeliveryoptimization-dev_1.0.0_arm64.deb

```

**Comment by jelalanne** [comment]
Did not work for me, still same error.

**Comment by jelalanne** [comment]
> > You might have a try with these 18.04 images :)
> > https://cdimage.ubuntu.com/releases/18.04/release/
> > But same problem as yours, I was looking for an answer too. Funny thing is that I figured this out once, but couldn't remember how I did it.
> 
> Thanks, @jelalanne for the Ubuntu 18.04 link. I installed it on a 2nd Raspberry and were able to successfully install the device update agent (with the same ansible script as with Ubuntu 20.04). However, Ubuntu 18.04 is kind of old and I'd rather prefer at least Ubuntu 20.04.

I am working on 20.04 so you might want to wait for us to find a solution ;)

**Comment by buentead** [comment]
> Did not work for me, still same error.

Same here. Also not working on Ubuntu 20.04.

**Comment by jelalanne** [comment]
Using do-client version 0.8.2 worked, but it's quite old. But it worked with installing deviceupdate-agent v0.8.2 too, not 1.0.1 which is the latest (annoyingly, its release file for arm64 is not available on the github release page). I remember now, that is what I was doing to overpass the problem last time.

`wget https://github.com/microsoft/do-client/releases/download/v0.8.2/ubuntu2004_arm64-packages.tar
tar -xvf ubuntu2004_arm64-packages.tar
sudo apt install -y ./deliveryoptimization-agent_0.6.0_arm64.deb ./libdeliveryoptimization_0.6.0_arm64.deb

wget https://github.com/Azure/iot-hub-device-update/releases/download/0.8.2/deviceupdate-agent_0.8.2_public_preview_arm64.deb
sudo apt install -y ./deviceupdate-agent_0.8.2_public_preview_arm64.deb`

NOTE: don't forget the relative path indicator './' before the file names, files might not be found otherwise.

IMPORTANT: @buentead just in case you didn't take that in account, installing ADU with IoT Edge is not enough. Don't forget to prepare the disk partitions to use A/B Update, install SWUpdate and everything. Sorry I don't have any tutorial, and that misses from MSFT documentation. I'm still trying to figure everything out.

@jw-msft Am I alright or did I say something wrong? Anything to advise?

**Comment by jw-msft** [comment]
Later today, I'll install arm64 ubuntu20.04 via RPi imager later and respond back on status of 1.0.1, which is recommended over 0.8.2

**Comment by jw-msft** [comment]
I tried out ubuntu 20.04.5 LTS on RPi3 via RaspberryPi Imager.

The problem is that 1.0.1 DU agent .deb package and AducIotAgent binary have an incorrect shared object dependency of libdeliveryoptimization.so.0.

Instead it should be libdeliveryoptimization.so.1  (NOTE: the .1 on the end).

For now, the workaround is to just a symlink to the .so.1 like this:
sudo ln -sf /usr/lib/libdeliveryoptimization.so.1.0.0 /usr/lib/libdeliveryoptimization.so.0

Here are the full steps I used to successfully install 1.0.1 DU agent:
1. Flash sd card using raspberry pi imager app https://downloads.raspberrypi.org/imager/imager_latest.exe
2. Login with ubuntu/ubuntu
3. Enter current password again for change password flow
4. Enter new password

```
$ cat /proc/cpuinfo | grep Model
Model           : Raspberry Pi 3 Model B Plus Rev 1.3

ubuntu@ubuntu:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

```
Install DO from release asset packages:
[Release v1.0.0 (2022-10-28) · microsoft/do-client (github.com)](https://github.com/microsoft/do-client/releases/tag/v1.0.0)

1. wget https://github.com/microsoft/do-client/releases/download/v1.0.0/ubuntu2004_arm64-packages.tar
2. tar xvf ubuntu2004_arm64-packages.tar
3. sudo apt-get install ./deliveryoptimization-agent_1.0.0_arm64.deb
4. sudo apt-get install ./libdeliveryoptimization_1.0.0_arm64.deb
5. wget http://packages.microsoft.com/ubuntu/20.04/prod/pool/main/d/deviceupdate-agent/deviceupdate-gent_1.0.1_ubuntu2004_arm64.deb
6. sudo apt-get update --fix-missing
7. sudo apt-get install libcurl4-openssl-dev
8. Workaround for missing libdeliveryoptimization.so.0 error:  sudo ln -sf /usr/lib/libdeliveryoptimization.so.1.0.0 /usr/lib/libdeliveryoptimization.so.0
9. sudo apt-get install ./deviceupdate-agent_1.0.1_ubuntu2004_arm64.deb
10. create a valid /etc/adu/du-config.json

Seems to be working for me:
```
2023-01-20T00:27:27.3786Z 3760[3760] [D] Log file created: /var/log/adu/du-agent.20230120-002727.log [zlog_init]
2023-01-20T00:27:27.3803Z 3760[3760] [I] Agent (linux; 1.0.1) starting. [main]
2023-01-20T00:27:27.3804Z 3760[3760] [I] Supported Update Manifest version: min: 4, max: 5 [main]
2023-01-20T00:27:27.3820Z 3760[3760] [I] Health check passed. [HealthCheck]
2023-01-20T00:27:27.3824Z 3760[3760] [I] Initializing PnP components. [ADUC_PnP_Components_Create]
2023-01-20T00:27:27.3824Z 3760[3760] [I] ADUC agent started. Using IoT Hub Client SDK 1.7.0 [AzureDeviceUpdateCoreInterface_Create]
2023-01-20T00:27:27.3824Z 3760[3760] [I] Calling ADUC_RegisterPlatformLayer [ADUC_MethodCall_Register]
2023-01-20T00:27:27.3827Z 3760[3760] [I] Loading extension 'Content Downloader'. Reg file : /var/lib/adu/extensions/content_downloader/extension.json [LoadExtensionLibrary]
2023-01-20T00:27:27.4055Z 3760[3760] [D] Determining contract version for content downloader. [LoadContentDownloaderLibrary]
2023-01-20T00:27:27.4056Z 3760[3760] [D] Got Contract 1.0 for content downloader [LoadContentDownloaderLibrary]
2023-01-20T00:27:27.4057Z 3760[3760] [I] Initializing command listener thread [InitializeCommandListenerThread]
2023-01-20T00:27:27.4059Z 3760[3760] [I] Command register at slot#0 [RegisterCommand]
2023-01-20T00:27:27.4059Z 3760[3760] [I] Agent running. [main]
2023-01-20T00:27:27.4061Z 3760[3761] [I] Command FIFO file created successfully. [TryCreateFIFOPipe]
2023-01-20T00:27:27.4063Z 3760[3760] [I] Attempting to create connection to IotHub using type: ADUC_ConnType_Device  [ADUC_DeviceClient_Create]
2023-01-20T00:27:27.4064Z 3760[3760] [I] IotHub Protocol: MQTT [GetIotHubProtocolFromConfig]
2023-01-20T00:27:27.4065Z 3760[3760] [I] IoTHub Device Twin callback registered. [ADUC_DeviceClient_Create]
2023-01-20T00:27:27.4066Z 3760[3760] [I] Refreshing the handle for the PnP channels. [ADUC_PnP_Components_HandleRefresh]
2023-01-20T00:27:27.4066Z 3760[3760] [I] Successfully re-authenticated the IoT Hub connection. [ADUC_Refresh_IotHub_Connection_SAS_Token]
```

**Comment by jelalanne** [comment]
Does this work for armhf architectures to?

**Comment by buentead** [comment]
@jw-msft : Works perfectly on my side as well. Thank you so much for the detailed workaround. Very helpful indeed.

Shall I keep the issue open until the new package will be available?

**Comment by jw-msft** [comment]
@jelalanne I know that Deb10 on armhf works. I will try it with Ubuntu 20.04 32-bit (based on Debian 11) on rpi3 and let you know.
@buentead Yes, let's keep this issue open until we've updated the package.
@nihemstr Please update here once package is updated.

**Comment by jw-msft** [comment]
@jelalanne  microsoft/do-client and DU Agent do not build armhf .deb, but installing the Debian 10 armhf .deb works on 32-bit ubuntu 20.04.5 LTS server armhf.

NOTE: it also needs the libdeliveryoptimization.so.0 symlink workaround and we'll need to check and update all the assets at this point.  //cc @nihemstr 

I was able to get it working with the following:
```sh
wget https://github.com/microsoft/do-client/releases/download/v1.0.0/debian10_arm32-packages.tar 
tar -xvf debian10_arm32-packages.tar
sudo apt-get install ./deliveryoptimization-agent_1.0.0_armhf.deb 
sudo apt-get install ./libdeliveryoptimization_1.0.0_armhf.deb
wget http://packages.microsoft.com/debian/10/prod/pool/main/d/deviceupdate-agent/deviceupdate-agent_1.0.0_debian10_armhf.deb
sudo ln -sf /usr/lib/libdeliveryoptimization.so.1.0.0 /usr/lib/libdeliveryoptimization.so.0  # Workaround
sudo apt-get install ./deviceupdate-agent_1.0.1_debian10_armhf.deb
```

**Comment by jelalanne** [comment]
Hello!
That didn't work for me...

![image](https://user-images.githubusercontent.com/32770824/217303389-7b0a0dc6-505d-4cf8-83e9-1a4242567eea.png)

![image](https://user-images.githubusercontent.com/32770824/217303468-6ac9dbe8-9e99-47df-8d26-84253fbd4c8a.png)

I'm using Ubuntu 20.04.5 preinstalled server from cdimage

Maybe a step I missed?

**Comment by jw-msft** [comment]
Issues have been resolved with [patch release 1.0.2 ](https://github.com/Azure/iot-hub-device-update/releases/tag/1.0.2) that was released today.

Please see this doc I created for steps to install the following from packages.microsoft.com after setting up /etc/apt/sources.list.d/:
https://github.com/Azure/iot-hub-device-update/blob/user/jw-msft/rpidoc/docs/agent-reference/how-to-install-deb-pkg-on-rpi.md
(which should eventually end up in develop branch [here](https://github.com/Azure/iot-hub-device-update/blob/develop/docs/agent-reference/how-to-install-deb-pkg-on-rpi.md) and in main [here](https://github.com/Azure/iot-hub-device-update/blob/rpidoc/docs/agent-reference/how-to-install-deb-pkg-on-rpi.md))


- ubuntu20.04+aarch64 on rpi3
- debian10+armhf on rpi3
- and ubuntu20.04+aarch64 on rpi4:

PR is here for the doc -> develop: #377
